event-sourcing/event-repository/event-repository-jdbc/src/main/java/uk/gov/justice/services/eventsourcing/repository/jdbc/event/EventJdbcRepository.java

```diff
@@ -198,14 +226,81 @@ protected String jndiName() throws NamingException {
-            final PreparedStatementWrapper ps = preparedStatementWrapperOf(SQL_PAGE);
+            final Object streamId = params.get(STREAM_ID);
+            if (offset == 0L && link.equals(HEAD)) {
+                return head(pageSize, streamId);
+            }
+            if (offset == 0L && link.equals(LAST)) {
+                return last(pageSize, streamId);
+            }
+            if (link.equals(PREVIOUS)) {
+                return forward(offset, pageSize, streamId);
+            }
+            if (link.equals(NEXT)) {
+                return backward(offset, pageSize, streamId);
+            }
+            return Stream.empty();
+        } catch (SQLException e) {
+            throw new JdbcRepositoryException(READING_STREAM_EXCEPTION, e);
+        }
+    }
+
+    private Stream<Event> last(final long pageSize, final Object streamId) throws SQLException {
+        final PreparedStatementWrapper ps;
+        ps = preparedStatementWrapperOf(LAST_RECORDS);
+        ps.setObject(1, streamId.toString());
+        ps.setLong(2, pageSize);
+        return streamOf(ps);
+    }
+
+    private Stream<Event> backward(final long offset, final long pageSize, Object streamId)
+            throws SQLException {
+        final PreparedStatementWrapper ps = preparedStatementWrapperOf(OLDER_RECORDS);
+        ps.setObject(1, streamId.toString());
+        ps.setLong(2, offset);
+        ps.setLong(3, pageSize);
+        return streamOf(ps);
+    }
+
+    private Stream<Event> forward(final long offset, final long pageSize, final Object streamId)
+            throws SQLException {
+        final PreparedStatementWrapper ps = preparedStatementWrapperOf(NEWER_RECORDS);
+        ps.setObject(1, streamId.toString());
+        ps.setLong(2, offset);
+        ps.setLong(3, pageSize);
+        return streamOf(ps).sorted(Comparator.comparing(Event::getSequenceId).reversed());
+    }
+
+    private Stream<Event> head(final long pageSize, final Object streamId) throws SQLException {
+        final PreparedStatementWrapper ps = preparedStatementWrapperOf(HEAD_RECORDS);
+        ps.setObject(1, streamId.toString());
+        ps.setLong(2, pageSize);
+        return streamOf(ps).sorted(Comparator.comparing(Event::getSequenceId).reversed());
+    }
+
+    @Override
+    public boolean recordExists(final long offset,
+                                final Link link,
+                                final long pageSize,
+                                final Map<String, Object> params) {
+        try {
+            final Object streamId = params.get(STREAM_ID);
+            PreparedStatementWrapper ps = preparedStatementWrapperOf(RECORD_EXISTS);
             ps.setObject(1, streamId.toString());
-            ps.setLong(2, pageSize);
-            ps.setLong(3, offset);
-            return streamOf(ps);
+            ps.setLong(2, offset);
+            try (ResultSet rs = ps.executeQuery()) {
+                rs.next();
+                if (rs.getInt(1) > 0) {
```

Comment:
rather than returning true if condition, can't you just return the value of the condition (remove the if)

Reply:
Fixed